File: apps/mobile/lib/services/order_math.dart
class OrderLineAmount {
  const OrderLineAmount({
    required this.unitPrice,
    required this.productQty,
  });

  final double unitPrice;
  final double productQty;
}

class OrderTotals {
  const OrderTotals({
    required this.lineTotal,
    required this.totalOrderValue,
  });

  final double lineTotal;
  final double totalOrderValue;
}

/// Single-line totals (legacy helper).
double calculateOrderTotal({
  required double unitPrice,
  required double productQty,
  required String discountType,
  required double discountValue,
}) {
  return calculateMultiLineOrderTotals(
    lines: [OrderLineAmount(unitPrice: unitPrice, productQty: productQty)],
    discountType: discountType,
    discountValue: discountValue,
  ).totalOrderValue;
}

/// Multi-line order totals: sum of line subtotals, then one order-level discount.
OrderTotals calculateMultiLineOrderTotals({
  required List<OrderLineAmount> lines,
  required String discountType,
  required double discountValue,
}) {
  if (lines.isEmpty) {
    throw ArgumentError('Order requires at least one line');
  }

  final lineTotal = _round(
    lines.fold<double>(
      0,
      (sum, line) => sum + _round(line.unitPrice * line.productQty),
    ),
  );

  if (discountType.toUpperCase() == 'PERCENTAGE') {
    return OrderTotals(
      lineTotal: lineTotal,
      totalOrderValue: _round(lineTotal * (1 - discountValue / 100)),
    );
  }

  return OrderTotals(
    lineTotal: lineTotal,
    totalOrderValue: _round((lineTotal - discountValue).clamp(0, double.infinity)),
  );
}

double _round(double value) => double.parse(value.toStringAsFixed(4));

const _paymentEpsilon = 0.00005;

/// Derive invoice collection status from paid amount vs amount due.
String deriveInvoiceStatusFromPayments({
  required double amountDue,
  required double paidAmount,
  required String billStatus,
}) {
  final total = amountDue;
  final paid = paidAmount;
  if (paid <= _paymentEpsilon) {
    return billStatus == 'SENT' ? 'SENT' : 'CREATED';
  }
  if (paid + _paymentEpsilon >= total) {
    return 'FULLY_PAID';
  }
  return 'PARTIALLY_PAID';
}

File: apps/api/src/orders/order-math.ts
type DiscountType = 'PERCENTAGE' | 'AMOUNT';

type OrderTotalsInput = {
  unitPrice: number;
  productQty: number;
  discountType: DiscountType;
  discountValue: number;
};

type OrderLineAmountInput = {
  unitPrice: number;
  productQty: number;
};

type OrderTotals = {
  lineTotal: number;
  totalOrderValue: number;
};

function roundMoney(value: number): number {
  return Math.round((value + Number.EPSILON) * 10000) / 10000;
}

export function lineSubtotal(unitPrice: number, productQty: number): number {
  if (unitPrice < 0 || productQty < 0) {
    throw new Error('Unit price and product qty must be non-negative');
  }
  return roundMoney(unitPrice * productQty);
}

/** Single-line totals (legacy helper). */
export function calculateOrderTotals(input: OrderTotalsInput): OrderTotals {
  return calculateMultiLineOrderTotals({
    lines: [{ unitPrice: input.unitPrice, productQty: input.productQty }],
    discountType: input.discountType,
    discountValue: input.discountValue,
  });
}

/**
 * Multi-line order totals: sum of line subtotals, then one order-level discount.
 */
export function calculateMultiLineOrderTotals(input: {
  lines: OrderLineAmountInput[];
  discountType: DiscountType;
  discountValue: number;
}): OrderTotals {
  if (!input.lines.length) {
    throw new Error('Order requires at least one line');
  }
  if (input.discountValue < 0) {
    throw new Error('Discount value must be non-negative');
  }

  const lineTotal = roundMoney(
    input.lines.reduce(
      (sum, line) => sum + lineSubtotal(line.unitPrice, line.productQty),
      0,
    ),
  );

  if (input.discountType === 'PERCENTAGE') {
    if (input.discountValue > 100) {
      throw new Error('Percentage discount cannot exceed 100');
    }
    const totalOrderValue = roundMoney(
      lineTotal * (1 - input.discountValue / 100),
    );
    return { lineTotal, totalOrderValue };
  }

  if (input.discountValue > lineTotal) {
    throw new Error('Discount amount cannot exceed line total');
  }
  const totalOrderValue = roundMoney(lineTotal - input.discountValue);
  return { lineTotal, totalOrderValue };
}

/**
 * Allocate order total across lines by pre-discount share
 * so product revenue sums to totalOrderValue.
 */
export function allocateLineRevenue(
  lineTotals: number[],
  totalOrderValue: number,
): number[] {
  const sum = roundMoney(lineTotals.reduce((a, b) => a + b, 0));
  if (lineTotals.length === 0) return [];
  if (sum <= 0) {
    return lineTotals.map(() => 0);
  }
  const allocated = lineTotals.map((lt) =>
    roundMoney((lt / sum) * totalOrderValue),
  );
  const allocatedSum = roundMoney(allocated.reduce((a, b) => a + b, 0));
  const drift = roundMoney(totalOrderValue - allocatedSum);
  if (drift !== 0 && allocated.length > 0) {
    allocated[allocated.length - 1] = roundMoney(
      allocated[allocated.length - 1] + drift,
    );
  }
  return allocated;
}
